tests/test_datastore.py

```diff
@@ -300,17 +299,145 @@ def testNestedTransaction(self):
+            """Can't ingest files outside of datastore root."""
+            with self.assertRaises(RuntimeError):
+                datastore.ingest(os.path.abspath(path), ref, transfer=None)
+            self.assertFalse(datastore.exists(ref))
+
+        def failNotImplemented(obj, path, ref):
+            with self.assertRaises(NotImplementedError):
+                datastore.ingest(path, ref, transfer=None)
+
+        if None in self.ingestTransferModes:
+            self.runIngestTest(failOutsideRoot)
+            self.runIngestTest(failInputDoesNotExist)
+            self.runIngestTest(succeed)
+        else:
+            self.runIngestTest(failNotImplemented)
+
+    def testIngestTransfer(self):
+        """Test ingesting existing files after transferring them.
+        """
+
+        for mode in ("copy", "move", "hardlink", "symlink"):
+            with self.subTest(mode=mode):
+                datastore = self.makeDatastore(mode)
+
+                def succeed(obj, path, ref):
+                    """Ingest a file by transferring it to the template location."""
+                    datastore.ingest(os.path.abspath(path), ref, transfer=mode)
+                    self.assertEqual(obj, datastore.get(ref))
+
+                def failInputDoesNotExist(obj, path, ref):
+                    """Can't ingest files if we're given a bad path."""
+                    with self.assertRaises(FileNotFoundError):
+                        datastore.ingest("this-file-does-not-exist.yaml", ref, transfer=mode)
+                    self.assertFalse(datastore.exists(ref))
+
+                def failOutputExists(obj, path, ref):
+                    """Can't ingest files if transfer destination already exists."""
+                    with self.assertRaises(FileExistsError):
+                        datastore.ingest(os.path.abspath(path), ref, transfer=mode)
+                    self.assertFalse(datastore.exists(ref))
+
+                def failNotImplemented(obj, path, ref):
+                    with self.assertRaises(NotImplementedError):
+                        datastore.ingest(os.path.abspath(path), ref, transfer=mode)
+
+                if mode in self.ingestTransferModes:
+                    self.runIngestTest(failInputDoesNotExist)
+                    self.runIngestTest(succeed, expectOutput=(mode != "move"))
+                    self.runIngestTest(failOutputExists)
+                else:
+                    self.runIngestTest(failNotImplemented)
+
 
 class PosixDatastoreTestCase(DatastoreTests, lsst.utils.tests.TestCase):
     """PosixDatastore specialization"""
     configFile = os.path.join(TESTDIR, "config/basic/butler.yaml")
     uriScheme = "file:"
+    ingestTransferModes = (None, "copy", "move", "hardlink", "symlink")
+
+    def setUp(self):
+        super().setUp()
+        self.root = tempfile.mkdtemp(dir=TESTDIR)
+        self.config = DatastoreConfig(self.configFile)
+        self.config["root"] = self.root
```

Comment:
I really think these should all be using `setConfigRoot` rather than looking inside the config here (it's what `setConfigRoot` is *for*. I'll make the change myself as part of my ticket since I'm rewriting `setConfigRoot` and this will be a good test for it (and with chained datastores added to these tests it will be even more important to be consistent).